### PR TITLE
Updating the roadmap and adding a design roadmap.

### PR DESCRIPTION
**Pull request by benvanik**
Mostly just a brain dump, but significantly better than nothing and something we can expand on as needed (or link to dedicated docs).

### CODE REVIEW

File: docs/roadmap_design.md [hunk 1]
```diff
@@ -0,0 +1,861 @@
...
+maximum concurrency (by having a very large ringbuffer) or maximum memory usage
+(by having a minimally small ringbuffer).
+
+Allocating tensors from the ringbuffer does not require sophisticated runtime
+packing as we can emit IR to calculate required sizes for dynamically shaped
+tensors. Whether a basic block reserves `%sz = constant 42 : index` bytes or
+`%sz = std.muli %cst, %dyn_dim : index` bytes doesn't materially change how the
+allocations are performed. Since almost all usage involves simple write head
+bumps there is no need for ahead-of-time memory planning or large fixed
+allocations, and since no buffer within the ringbuffer can alias we can have
```
Comment (stellaraccident) [inline docs/roadmap_design.md]: Comma after "alias"

File: docs/roadmap_design.md [hunk 2]
```diff
@@ -0,0 +1,861 @@
...
+
+Waking the coroutines can be performed by either an application-provided
+callback in the case of the application already having a periodic event which is
+doing bookkeeping (such as frame end callbacks when rendering or Looper idle
+events on Android), giving direct control over the frequency and location which
+IREE utilizes to perform additional work. A helper will be provided as well that
+runs a dedicated IREE thread to do this, but the expectation is that
+applications can often do a better (and importantly more predictable) job.
+
+By utilizing coroutines IREE will have a way to fill traditional pipeline
```
Comment (stellaraccident) [inline docs/roadmap_design.md]: Comma after "coroutines"

File: docs/roadmap_design.md [hunk 3]
```diff
@@ -0,0 +1,861 @@
...
+```mlir
+// submit work
+hal.device.yield %semaphore4 >= %sem4_target, %semaphore5 >= %sem5_target
+// continue here, possibly much later in time
+```
+
+#### Cellular Batching
+<a id="markdown-cellular-batching" name="cellular-batching"></a>
+
+Though coroutines help throughput there is a way we've found to reduce latency
```
Comment (stellaraccident) [inline docs/roadmap_design.md]: Comma after "throughput"

File: docs/roadmap_design.md [hunk 4]
```diff
@@ -0,0 +1,861 @@
...
+
+There are many cases to explore and as cellular batching can have performance
+benefits of several orders of magnitudes it'll be one of the primary areas of
+research in the long-term.
+
+### Lowering to LLVM IR
+<a id="markdown-lowering-to-llvm-ir" name="lowering-to-llvm-ir"></a>
+
+For scenarios where dynamic module loading is not required and entire modules
+can be compiled into applications we can lower the VM IR to LLVM IR within
```
Comment (stellaraccident) [inline docs/roadmap_design.md]: Comma after "applications"

File: docs/roadmap_design.md (hunk 2 above)
Comment (benvanik) [reply docs/roadmap_design.md]: (that'd be a comma splice)

File: docs/roadmap_design.md (hunk 1 above)
Comment (benvanik) [reply docs/roadmap_design.md]: (that'd be a comma splice)

**Pull request closed without merging**